### PR TITLE
refactor cli flags. unmarshal config file using yaml package

### PR DESCRIPTION
**Pull request by jacobweinstock**
refactor cli flags. unmarshal config file using yaml package instead of viper

### CODE REVIEW

File: cmd/deploy.go [hunk 1]
```diff
@@ -10,7 +10,8 @@ import (
 	"path/filepath"
 	"time"
 
-	"github.com/spf13/viper"
+	//"github.com/spf13/viper"
```
Comment (j-griffith) [inline cmd/deploy.go:13]: We can just delete this

File: cmd/deploy.go [hunk 2]
```diff
@@ -109,9 +106,10 @@ func runProvider() {
 	var bootstrap provider.Bootstrapper
 	if deploymentType == "capv" {
 		vsProvider := new(vsphere.MgmtBootstrapCAPV)
-		errJ := viper.Unmarshal(&vsProvider)
+		errJ := yaml.Unmarshal(specContents, &vsProvider)
+		//errJ := viper.Unmarshal(&vsProvider)
```
Comment (j-griffith) [inline cmd/deploy.go:110]: same

File: cmd/deploy.go [hunk 3]
```diff
@@ -120,9 +118,10 @@ func runProvider() {
 		bootstrap = vsProvider
 	} else if deploymentType == "rke" {
 		vsProvider := new(vsphere.MgmtBootstrapRKE)
-		errJ := viper.Unmarshal(&vsProvider)
+		errJ := yaml.Unmarshal(specContents, &vsProvider)
+		//errJ := viper.Unmarshal(&vsProvider)
```
Comment (j-griffith) [inline cmd/deploy.go:122]: remove commented line

File: cmd/deploy.go [hunk 4]
```diff
@@ -177,9 +176,10 @@ func runEngine() {
 
 	if deploymentType == "capv" {
 		engine := capv.MgmtCluster{}
-		errJ := viper.Unmarshal(&engine)
+		errJ := yaml.Unmarshal(specContents, &engine)
+		//errJ := viper.Unmarshal(&engine)
```
Comment (j-griffith) [inline cmd/deploy.go:180]: ditto

File: cmd/deploy.go [hunk 5]
```diff
@@ -194,9 +194,10 @@ func runEngine() {
 		rkeDockerEnv := os.Getenv("CAKE_RKE_DOCKER")
 		if rkeDockerEnv != "" {
 			engine := rke.NewMgmtClusterFullConfig()
-			errJ := viper.Unmarshal(&engine)
+			errJ := yaml.Unmarshal(specContents, &engine)
+			//errJ := viper.Unmarshal(&engine)
```
Comment (j-griffith) [inline cmd/deploy.go:198]: here too

File: cmd/deploy.go [hunk 6]
```diff
@@ -206,9 +207,10 @@ func runEngine() {
 			engineName = engine
 		} else {
 			engine := rkecli.NewMgmtClusterCli()
-			errJ := viper.Unmarshal(&engine)
+			errJ := yaml.Unmarshal(specContents, &engine)
+			//errJ := viper.Unmarshal(&engine)
```
Comment (j-griffith) [inline cmd/deploy.go:211]: one more

File: cmd/deploy.go [hunk 7]
```diff
@@ -206,9 +207,10 @@ func runEngine() {
...
 			if errJ != nil {
-				log.Fatalf("unable to decode into struct, %v", errJ.Error())
+				log.Fatalf("unable to parse config, %v", errJ.Error())
```
Comment (j-griffith) [inline cmd/deploy.go:213]: unpopular opinion, throw in the name of the file we were trying to parse here?  Or better yet, output it in the logs somewhere above; BTW, this is another use for the folder approach that I thought of; maybe we want to create a log file in the cluster-name directory for future use.

Comment (jacobweinstock) [reply cmd/deploy.go:213]: sounds good. i'll add the config file path in the message.

File: cmd/deploy.go (hunk 6 above)
Comment (jacobweinstock) [reply cmd/deploy.go:211]: updated

File: cmd/deploy.go (hunk 5 above)
Comment (jacobweinstock) [reply cmd/deploy.go:198]: updated

File: cmd/deploy.go (hunk 4 above)
Comment (jacobweinstock) [reply cmd/deploy.go:180]: updated

File: cmd/deploy.go (hunk 3 above)
Comment (jacobweinstock) [reply cmd/deploy.go:122]: updated

File: cmd/deploy.go (hunk 2 above)
Comment (jacobweinstock) [reply cmd/deploy.go:110]: updated

File: cmd/deploy.go (hunk 1 above)
Comment (jacobweinstock) [reply cmd/deploy.go:13]: updated

File: cmd/genconfig.go [hunk 8]
```diff
@@ -67,20 +62,27 @@ func fileExists(fn string) bool {
...
+		os.MkdirAll(clusterNamePath, 0700)
 	}
-	clusterSpec = fmt.Sprintf("%s/spec.yaml", specPath)
-	if fileExists(clusterSpec) {
-		fmt.Println("A cluster spec file already exists for the cluster-name specified, please use another name or delete the existing spec.yml file")
+	if specFile == "" {
+		specFile = defaultSpecDir + defaultSpecFileName
+	}
+	clusterSpec = specFile
+	if !fileExists(specFile) {
```
Comment (j-griffith) [inline cmd/genconfig.go:77]: This call is going to fail, the part you changed was a safety to make sure we did NOT have a cluster spec already and accidentally overwrite it. You change the logic here from `exists` to `! exists`, the result is it's never going to be there in a normal run and fail on the lines below and exit.

It also means you have the opposite behavior, if you ever specify something that already exists inadvertantly then you're going to be trying to overwrite it.

Comment (jacobweinstock) [reply cmd/genconfig.go]: fixed

**Comment by jacobweinstock** [comment]
> It looks like this is a combination of reverting the cluster-name folder approach (but keeping the option there) and going back to ~/.cake/cake.yaml
> 
> If you want to stick with only one existence of a config `~/.cake/cake.yaml` that's fine with me, you'll just need to add logic to check and archive if one already exists.
> 
> What we had before was requiring a cluster-name for deploy and destroy. I'm fine if you just want to go back to the first method, but this doesn't actually do that right now either (see notes about genconfig). Even if I were to specify a cluster-name or path it's going to be ignored.

@j-griffith I fixed `genconfig` subcommand (although its not useful at the moment because its not creating a proper spec file) and updated deploy subcommand as well. I put the default spec file name back to `spec.yaml`.

The initial push of this PR was definitely premature, sorry about that. I was just trying to get `deploy` working. I do like the cluster-name folder approach. That's back in. For UX purposes the `deploy` subcommand needs a flag to specify a spec file. If that flag doesn't exist then the user would have to create a directory with the cluster name (ie ~/.cake/<cluster-name>/) and then put a `spec.yaml` file in there. Not clear or intuitive, in my opinion. 

Sorry for all the confusion and churn on this. Let me know if this will work.

**Comment by j-griffith** [comment]
Thanks @jacobweinstock, so this fixes things up; it looks like your approach here is that genconfig can take a name, but will always write out to `$HOME/.cake/`, that seems fine; the problem is that the changes to use `-f` in subsequent commands like deploy don't work.  

I'm fine if we want to just remove the -f option from those for now (in other words everything is always under `$HOME/.cake` and we can enhance that later if we want.  So what I'm getting at here is with your PR, just remove that option altogether, offer the name option only and we can go from there once thins are stabilized and we get some more mileage on things.

**Comment by jacobweinstock** [comment]
@j-griffith sorry, im not following. genconfig writes the spec to ~/.cake/\<cluster-name\>/spec.yaml 

image


and the deploy subcommand, as far as i can tell, is working just fine with the `-f` flag. 

image

**Comment by j-griffith** [comment]
> @j-griffith sorry, im not following. genconfig writes the spec to ~/.cake/<cluster-name>/spec.yaml
> 
> image
> 
> and the deploy subcommand, as far as i can tell, is working just fine with the `-f` flag.
> 
> image

Right, but what's the `-f` option in deploy for?  If for example I create something like `~/alt-cake/stupid-cluster/spec.yaml` and then try and specify that from deploy using `... -n stupid-cluster -f ~/alt-cake` it looks like it's still going to just create a random cluster-name and try to use that?  

Look at your log output on the second screen shot, it's using a cluster name "/root/.cake/driving-turtle", and I suspect it's failing because there's no spec file there (but it's not very clear in the error message if that's the case or not.

**Comment by jacobweinstock** [comment]
> > @j-griffith sorry, im not following. genconfig writes the spec to ~/.cake//spec.yaml
> > image
> > and the deploy subcommand, as far as i can tell, is working just fine with the `-f` flag.
> > image
> 
> Right, but what's the `-f` option in deploy for? If for example I create something like `~/alt-cake/stupid-cluster/spec.yaml` and then try and specify that from deploy using `... -n stupid-cluster -f ~/alt-cake` it looks like it's still going to just create a random cluster-name and try to use that?
> 
> Look at your log output on the second screen shot, it's using a cluster name "/root/.cake/driving-turtle", and I suspect it's failing because there's no spec file there (but it's not very clear in the error message if that's the case or not.

`-f` is to override the default spec file location, which is `~/.cake/<cluster-name>/spec.yaml`. with deploy if you specify `-n` and `-f`, if the cluster name exists it wont be created. if it doesnt exist it will be created. and since the `-f` is being specified cake wont look for a config at `~/.cake/<cluster-name>/spec.yaml` it will use the spec from the `-f` flag. 

the 2nd image is not using a `-n` flag so a randomly created cluster-name directory will be created. the `-f` flag is specified, so deploy will use the file specified for the spec and not look in the default location (`~/.cake/<cluster-name>/spec.yaml`)

**Comment by j-griffith** [comment]
Ok, I see; you're doing a different option here; rather than the path option to change the default working directory, you're just allowing users to point to a config spec and still using the .cake/cluster-name for any output artifacts.  That's cool, probably a better option that what I had there before anyway to be honest.

I'll probably add something to the dos later to explain all of this clearly, but for now I'm down with this.

**Pull request merged**